Lib/bdb.py

```diff
@@ -378,7 +378,7 @@ def set_trace(self, frame=None):
             # We need f_trace_lines == True for the debugger to work
             frame.f_trace_lines = True
             frame = frame.f_back
-        self.set_step()
+        self.set_stepinstr()
```

Comment:
So, if I understand correctly, this will turn on opcode tracing for all frames above the entry frame? And this will affect all `Bdb` subclasses? Seems like it could hurt perf for those tools, since it's flip flopping the instrumentation version on those code objects (right?).

I'm wondering if it would be more performant to avoid the whole opcode tracing dance unless some condition is met, like the `user_opcode` function is overridden, or some attribute like `self.enable_opcode_tracing` is set.

Reply:
Yes this will turn on opcode tracing for all frames, but this is a one time per debugger entry thing (breakpoints/until/next etc. are not affected). Are you worried about the boot time of the debugger?

First of all I don't think there are a lot of tools depending on `Bdb` (but not `Pdb`) in the market - not that we don't need to care about them at all. And I think it's consistent to make `Bdb.set_trace` stop immediately so this opcode tracing should be needed for all `Bdb` subclasses.

I'm kind of confused about which performance you have doubts about - starting from the first control command (s/n/r/c), the debugger should run at the normal speed. So there are probably two extra instrumentations (enable/disable opcode tracing) performed, but they seem like necessary.

Reply:
I guess I was concerned about what happens once the higher-up frames resume execution. But I guess they also have their line events toggled today, so they'll need to stop and figure out if they need to be instrumented either way.

If you're not worried, then I'm not... I mostly wanted to make sure we weren't introducing unnecessary overhead. But we should probably update the docs for `Bdb` (and call it out in the changelog) since client behavior will be affected (even if for the better).

Reply:
The higher-up frames will be instrumented when pdb does `sys.settrace(trace)` and will be de-instrumented when pdb does `sys.settrace(None)` - that was something that's already there. `sys.settrace()` will instrument all the executing code objects - no lazy instrumentation there. So once pdb is brought up, we already need to instrument them all. And yes, toggling `f_trace_opcode` will instrument them again.

From my perspective, the boot time of pdb does not matter that much as long as it's still in ms range. It's an interactive tool and human's time resolution is pretty bad.

Reply:
Yeah, I was looking at it from the perspective of throwing away all JIT code for the calling frames. But as you note, we already do that. So I'm fine with this. Let's just call out `Bdb` in the docs changes.